### PR TITLE
[fea] Expose the arena mr to the Python interface.

### PR DESCRIPTION
**Pull request by trivialfis**
## Description

Close https://github.com/rapidsai/rmm/issues/830 .

- Add the arena allocator to the public Python interface.
- Small changes to the logger initialization to avoid exposing spdlog in the shared objects.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: include/rmm/mr/device/arena_memory_resource.hpp [hunk 1]
```diff
@@ -97,7 +97,10 @@ class arena_memory_resource final : public device_memory_resource {
     : global_arena_{upstream_mr, arena_size}, dump_log_on_failure_{dump_log_on_failure}
   {
     if (dump_log_on_failure_) {
-      logger_ = spdlog::basic_logger_mt("arena_memory_dump", "rmm_arena_memory_dump.log");
+      logger_ =
+        std::make_shared<spdlog::logger>("arena_memory_dump",
+                                         std::make_shared<spdlog::sinks::basic_file_sink_mt>(
```
Comment (wence-) [inline include/rmm/mr/device/arena_memory_resource.hpp:102]: *question*: Why this change?

Comment (trivialfis) [reply include/rmm/mr/device/arena_memory_resource.hpp:102]: It's changed to avoid exposing a spdlog symbol to the ABI. I assume the basic_logger_mt uses a global map to store loggers with default visibility.
The truncate file is the same as the logging resource, just following the convention there.

Comment (wence-) [reply include/rmm/mr/device/arena_memory_resource.hpp:102]: Does `spdlog::basic_logger_mt` not just return a `shared_ptr<spdlog::logger>`? Which symbol are we now not exposing in the ABI?

Comment (trivialfis) [reply include/rmm/mr/device/arena_memory_resource.hpp:102]: Please see this test failure: https://github.com/rapidsai/rmm/actions/runs/11532359492/job/32104141080

Comment (wence-) [reply include/rmm/mr/device/arena_memory_resource.hpp:102]: Ah, thanks.

File: include/rmm/mr/device/detail/arena.hpp [hunk 2]
```diff
@@ -39,9 +39,7 @@
 #include <optional>
 #include <set>
 
-namespace RMM_NAMESPACE {
-namespace mr::detail::arena {
-
+namespace rmm::mr::detail::arena {
```
Comment (wence-) [inline include/rmm/mr/device/detail/arena.hpp:42]: This changes the `__attribute__((visibility))` of this namespace from `"default"` to `"hidden"`. Is this deliberate, or an inadvertent editor-induced change?

File: python/rmm/rmm/pylibrmm/memory_resource.pyx [hunk 3]
```diff
@@ -310,6 +311,49 @@ cdef class PoolMemoryResource(UpstreamResourceAdaptor):
...
+        Parameters
+        ----------
+        upstream_mr : DeviceMemoryResource
+            The DeviceMemoryResource from which to allocate memory for arenas.
+        arena_size : int, optional
+            Size in bytes of the global arena. Defaults to half of the available memory
+            on the current device.
+        dump_log_on_failure : bool, optional
+            Whether to dump the arena on allocation failure.
+
```
Comment (wence-) [inline python/rmm/rmm/pylibrmm/memory_resource.pyx:352]:
```suggestion
            Whether to dump the arena on allocation failure.
```

File: include/rmm/mr/device/detail/arena.hpp (hunk 2 above)
Comment (trivialfis) [reply include/rmm/mr/device/detail/arena.hpp:42]: It's deliberate.

File: python/rmm/rmm/pylibrmm/memory_resource.pyx (hunk 3 above)
Comment (trivialfis) [reply python/rmm/rmm/pylibrmm/memory_resource.pyx:352]: GitHub shows:
```
This suggestion is invalid because no changes were made to the code
```

File: include/rmm/mr/device/detail/arena.hpp (hunk 2 above)
Comment (wence-) [reply include/rmm/mr/device/detail/arena.hpp:42]: Ah, OK, can we just hide that one symbol? Add `RMM_HIDDEN` for that symbol.

See discussion in https://github.com/rapidsai/rmm/issues/1652, particularly, we need to be careful about visibility of objects that might make it in to the public namespace.

File: python/rmm/rmm/pylibrmm/memory_resource.pyx (hunk 3 above)
Comment (wence-) [reply python/rmm/rmm/pylibrmm/memory_resource.pyx:352]: Ah, I meant to mark deleting the empty line 352.

So:
```
        Whether to dump the arena on allocation failure.
    """
```

File: include/rmm/mr/device/detail/arena.hpp (hunk 2 above)
Comment (trivialfis) [reply include/rmm/mr/device/detail/arena.hpp:42]: Thank you for sharing. I will check whether the exception can be thrown from there. Considering that the namespace is `detail` and is only used by the MR, which is exposed in the public, it should be fine. That said, I will double check.
Who doesn't love c++ linking, so much fun.

Comment (trivialfis) [reply include/rmm/mr/device/detail/arena.hpp:42]: @wence- Did a quick test with XGBoost using rmm by inserting a `RMM_EXPECT(false, "test")` in allocate, the exception can go through the shared objects.
```
    _check_call(ret)
  File "/home/jiamingy/workspace/xgboost_dev/xgboost/python-package/xgboost/core.py", line 299, in _check_call
    raise XGBoostError(py_str(_LIB.XGBGetLastError()))
xgboost.core.XGBoostError: [08:30:53] /home/jiamingy/workspace/xgboost_dev/xgboost/src/common/device_vector.cu:23: Memory allocation error on worker 0: RMM failure at: /home/jiamingy/.anaconda/envs/xgboost_dev_125/include/rmm/mr/device/detail/arena.hpp:588: test
```

File: python/rmm/rmm/pylibrmm/memory_resource.pyx (hunk 3 above)
Comment (trivialfis) [reply python/rmm/rmm/pylibrmm/memory_resource.pyx]: Done.

File: include/rmm/mr/device/detail/arena.hpp (hunk 2 above)
Comment (harrism) [reply include/rmm/mr/device/detail/arena.hpp:42]: I am confused. Have we just exposed the whole file, or can you just unhide parts of it?

Comment (trivialfis) [reply include/rmm/mr/device/detail/arena.hpp:42]: We have exposed the memory resource but not the arena. As long as the exception goes through the memory resource them it should go through the shared object.

Comment (trivialfis) [reply include/rmm/mr/device/detail/arena.hpp:42]: The visibility is controlled by namespace instead of translation unit.

Comment (harrism) [reply include/rmm/mr/device/detail/arena.hpp:42]: oh I had it backwards. I thought this change was exposing more, but it's actually hiding more.

Comment (wence-) [reply include/rmm/mr/device/detail/arena.hpp:42]: May I suggest that we make the hiding explicit?

```
namespace RMM_HIDDEN rmm::mr::detail::arena {
```
?

Comment (trivialfis) [reply include/rmm/mr/device/detail/arena.hpp:42]: I'm totally happy to make the change. But do you want to do it for the entire repository or do you want me to make the change to this specific file? Or maybe the CMake hidden visibility option is sufficient?

Comment (trivialfis) [reply include/rmm/mr/device/detail/arena.hpp]: Switched to hiding only the spdlog function as suggested offline.

**Review by wence- — APPROVED** [review]

**Comment by wence-** [comment]
Thanks!
